File: Java/DSA/Basics/BasicHashOperations.java
import java.util.*;
import java.util.Map.Entry;

public class BasicHashOperations {
    public static void main(String[] args) {
        int[] arr = {1,2,4,5,3,5,3,6,6,4,6,8,4,5,6,5,3};
        Map<Integer,Integer> frequencyMap = frequencyCounter(arr);
        System.out.println(frequencyMap.toString());

        List<Integer> maxFrequencyElements = maxFrequencyElements(arr);
        System.out.println(maxFrequencyElements.toString());
    }

    static Map<Integer,Integer> frequencyCounter(int[] arr){
        Map<Integer,Integer> frequencyMap = new HashMap<>();

        for (int n : arr){
            frequencyMap.put(n,frequencyMap.getOrDefault(n,0)+1);
        }

        return frequencyMap;
    }

    static ArrayList<Integer> maxFrequencyElements(int[] arr){
        Map<Integer,Integer> frequencyMap = frequencyCounter(arr);
        int max = Collections.max(frequencyMap.values());
        ArrayList<Integer> res = new ArrayList<>();
        for (Entry<Integer,Integer> entry : frequencyMap.entrySet()){
            if (entry.getValue() == max){
                res.add(entry.getKey());
            }
        }
        return res;
    }
}

File: Python/DSA/Basics/basic_hash_operations.py
def count_frequency(arr):
    frequency_map = {}

    for num in arr:
        if num in frequency_map:
            frequency_map[num] +=1
        else:
            frequency_map[num] = 1
    
    return frequency_map


def max_frequency_elements(arr):
    frequency_map = count_frequency(arr)
    max_frequency = max(frequency_map.values())

    high_freq_elements = [key for key, value in frequency_map.items() if value == max_frequency]

    return high_freq_elements


arr = [1,3,5,7,4,6,3,4,5,4,6,8,6,5,3,5,3,5,4,7,4]
print(count_frequency(arr))
print(max_frequency_elements(arr))